### PR TITLE
UNPLUGGED-81 | Added a way to launch feed reprocessing after changing the settings

### LINKED ISSUE #81: Prestashop - reindex when a setting has changed
As a shop owner when I do a change of any setting and it requires to index the search engines again I want to be able to process the index from the platform.

Some settings changes require to process again the indice and currently the user has to do it manually from doofinder admin to see the changes effective in the layer.

### Definition:
- When a setting change requires to index the search engines, then add a button in the message to allow to launch the indexing process and show a message to inform that the indice will be processed again: 
"PARA QUE LOS CAMBIOS TENGAN EFECTO, RECUERDA QUE DEBES REPROCESAR EL INCIDE" [Lanzar indexación]"
- Show section when indice is being processed**

- Remove current message:
"SI HAS CAMBIADO ALGO EN LAS OPCIONES DE EXPORTACIÓN DE TU FEED, RECUERDA QUE DEBES REPROCESARLO"

### PR DESCRIPTION
**Pull request by davidmolinacano**
Required by: https://github.com/doofinder/doofinder-prestashop/issues/81 and cannot be deployed before this one is deployed: https://github.com/doofinder/doomanager/pull/4522.  

This issue consists on two main parts: one is included in the other PR, which involves the creation of an endpoint in GraphQL on DooManager and the other part is this one which consists on calling the endpoint after clicking the "launch reindexing" button in Prestashop.

prestashop-reindex

IMPORTANT: @CristobalDolz and the support team have to test the plugin before deploying it to the general public

### CODE REVIEW

File: doofinder.php [hunk 1]
```diff
@@ -1995,4 +2001,33 @@ private function uninstallTabs()
 
         return $tab->delete();
     }
+
+    private function invokeReindexing()
```
Comment (frobs) [inline doofinder.php]: Cuando nos tenemos que comunicar con nuestra API solemos sacar las peticiones fuera del archivo principal por ejemplo doofinder-prestashop/lib/doofinder_api_items.php.

Quizás podríamos sacar la lógica de la petición a otra clase y así no lo tenemos en el archivo principal, veo que checkOutsideConnection y createStore también tienen las peticiones en los propios métodos, no se si podrían sacarse también.
¿Tu que opinas @ivan-doofinder ?

Comment (davidmolinacano) [reply doofinder.php]: Por mi parte, opino que tienes toda la razón, creo que la idea sería dejar el doofinder.php principal lo más limpito posible y que el procesamiento "intenso" se haga en otro archivo. De todos modos, quedo a la espera de que @ivan-doofinder nos confirme tu 2ª cuestión. Gracias por el feedback, @frobs

Comment (davidmolinacano) [reply doofinder.php]: Lo he cambiado aquí: https://github.com/doofinder/doofinder-prestashop/pull/143/commits/ed970c69d097338767ed2f7efb6fab41d435aa0d

File: doofinder.php [hunk 2]
```diff
@@ -35,15 +35,15 @@ class Doofinder extends Module
     const DOOMANAGER_URL = 'https://admin.doofinder.com';
     const GS_SHORT_DESCRIPTION = 1;
     const GS_LONG_DESCRIPTION = 2;
-    const VERSION = '4.6.5';
+    const VERSION = '4.7.0';
```
Comment (ivan-doofinder) [inline doofinder.php]:
```suggestion
    const VERSION = '4.6.6';
```

File: doofinder.php [hunk 3]
```diff
@@ -35,15 +35,15 @@ class Doofinder extends Module
...
     const YES = 1;
     const NO = 0;
 
     public function __construct()
     {
         $this->name = 'doofinder';
         $this->tab = 'search_filter';
-        $this->version = '4.6.5';
+        $this->version = '4.7.0';
```
Comment (ivan-doofinder) [inline doofinder.php]:
```suggestion
        $this->version = '4.6.6';
```

File: lib/doofinder_api_items.php [hunk 4]
```diff
@@ -55,6 +60,24 @@ public function deleteBulk($payload)
         return $this->post($url, $payload);
     }
 
+    /**
+     * Make a request to the API to reprocess all the feeds
+     *
+     * @param array items ids
+     */
+    public function invokeReindexing()
```
Comment (ivan-doofinder) [inline lib/doofinder_api_items.php]: This class (DoofinderApiItems) is for update on save. Please put it in another place. If necessary create a new class (For example: DoofinderApiIndex).

File: lib/doofinder_api_items.php [hunk 5]
```diff
@@ -55,6 +60,24 @@ public function deleteBulk($payload)
...
+    {
+        $installation_id = Configuration::get('DF_INSTALLATION_ID');
+        $json_data = json_encode(['query' => 'mutation { process_store_feeds(id: "' . $installation_id . '", callback_url: "' . $this->getProcessCallbackUrl() . '") { id }}']);
```
Comment (ivan-doofinder) [inline lib/doofinder_api_items.php]: installation_id must be a parameter of the function.
Also callback_url

File: lib/doofinder_api_items.php [hunk 6]
```diff
@@ -55,6 +60,24 @@ public function deleteBulk($payload)
...
+        $response = $this->post($this->api_url . '/api/v1/graphql.json', $json_data);
+        if (empty($response)) {
+            return;
+        }
+        if (empty($response_array['errors'])) {
+            Configuration::updateValue('DF_FEED_INDEXED', false);
+        }
```
Comment (ivan-doofinder) [inline lib/doofinder_api_items.php]: This logic has to be out.
This class should be for connecting to the API only.

File: lib/doofinder_api_items.php [hunk 7]
```diff
@@ -70,4 +93,14 @@ private function post($url, $payload)
...
+
+    /**
+     * Get Process Callback URL
+     *
+     * @return string
+     */
+    private function getProcessCallbackUrl()
+    {
+        return Context::getContext()->link->getModuleLink('doofinder', 'callback', []);
+    }
```
Comment (ivan-doofinder) [inline lib/doofinder_api_items.php]: also, out

**Review by ivan-doofinder — CHANGES_REQUESTED** [review]

File: lib/doofinder_api_items.php (hunk 7 above)
Comment (davidmolinacano) [reply lib/doofinder_api_items.php]: Done on https://github.com/doofinder/doofinder-prestashop/pull/143/commits/13a4e4631a3fff59967320600ca685099550cd59. Thank you for the feedback! :smile:

File: lib/doofinder_api_items.php (hunk 6 above)
Comment (davidmolinacano) [reply lib/doofinder_api_items.php]: Totally agree, changed on: https://github.com/doofinder/doofinder-prestashop/pull/143/commits/13a4e4631a3fff59967320600ca685099550cd59

File: lib/doofinder_api_items.php (hunk 5 above)
Comment (davidmolinacano) [reply lib/doofinder_api_items.php]: Changed on: https://github.com/doofinder/doofinder-prestashop/pull/143/commits/13a4e4631a3fff59967320600ca685099550cd59

File: lib/doofinder_api_items.php (hunk 4 above)
Comment (davidmolinacano) [reply lib/doofinder_api_items.php]: Created new class on: https://github.com/doofinder/doofinder-prestashop/pull/143/commits/13a4e4631a3fff59967320600ca685099550cd59

File: doofinder.php (hunk 3 above)
Comment (davidmolinacano) [reply doofinder.php]: Ok, I misunderstood this part, changed to patch version: https://github.com/doofinder/doofinder-prestashop/pull/143/commits/13a4e4631a3fff59967320600ca685099550cd59

File: doofinder.php [hunk 8]
```diff
@@ -1250,6 +1255,28 @@ private function deleteItemsApi($hashid, $type, $payload)
         }
     }
 
+    private function indexApiInvokeReindexing()
+    {
+        require_once dirname(__FILE__) . '/lib/doofinder_api_index.php';
+
+        $region = Configuration::get('DF_REGION');
+        $api_key = $region . '-' . Configuration::get('DF_AI_APIKEY');
```
Comment (ivan-doofinder) [inline doofinder.php]:
```suggestion
        $api_key = Configuration::get('DF_API_KEY');
```
Use DF_API_KEY.
DF_AI_APIKEY is for autoinstaller only

File: lib/doofinder_api_items.php [hunk 9]
```diff
@@ -19,6 +19,11 @@
 
 class DoofinderApiItems
 {
+    private $hashid;
+    private $api_key;
+    private $api_url;
+    private $type;
+
```
Comment (ivan-doofinder) [inline lib/doofinder_api_items.php]: Delete this. In this file you do not have to make changes

**Review by ivan-doofinder — CHANGES_REQUESTED** [review]

File: lib/doofinder_api_items.php (hunk 9 above)
Comment (davidmolinacano) [reply lib/doofinder_api_items.php]: I know that this file is unrelated, but I applied the boy-scout rule and I tried to fix them, but I'm going to revert them if you consider it's better to apply them on a different PR :smiley:

File: doofinder.php (hunk 8 above)
Comment (davidmolinacano) [reply doofinder.php]: Changed on https://github.com/doofinder/doofinder-prestashop/pull/143/commits/f0c445600bbddbc8f382b7306a50c454ad70070a

**Review by ivan-doofinder — APPROVED** [review]

**Review by frobs — APPROVED** [review]

**Pull request merged**